mtasa-resources/CODING_GUIDELINES.md

```diff
@@ -49,42 +59,92 @@ function exampleFunction(value)
 
-Ensure your code editor (e.g. [Visual Studio Code](https://code.visualstudio.com/) 
-applies the rules established by the project's **.editorconfig** file.
+Rather than using anonymous functions (i.e when calling [`addEventHandler`](https://wiki.multitheftauto.com/wiki/AddEventHandler)), use a named local function instead. For example:
 
-# General principles
+```lua
+-- Rather than this:
+addEventHandler("onPlayerConnect", root, function()
+    outputChatBox("Hello!")
+end)
+
+-- Do this:
+local function onConnect()
+    outputChatBox("Hello!")
+end
+addEventHandler("onPlayerConnect", root, onConnect)
+```
 
--   Write clear, readable, and maintainable code.
--   Follow the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
-    (Don't Repeat Yourself) principle.
--   Adhere to the [KISS](https://en.wikipedia.org/wiki/KISS_principle)
-    (Keep It Simple, Stupid) principle.
--   Use meaningful variable and function names that convey their purpose.
--   Comment your code where necessary to explain the logic.
+### Consistent naming conventions
 
-# Script security
+All function names and variables should use the camel case naming convention. Constant variables should use upper snake case (see below).
 
-Follow the [Script security](https://wiki.multitheftauto.com/wiki/Script_security) 
-principles established for MTA:SA scripts to ensure the safety and integrity of your code.
+### Use of constants and avoiding [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming))
+
+Don't use magic numbers in blocks of code - instead, define such values as "constants" at the top of the file using upper snake case.
+
+*Note: the version of Lua used by MTA does not support the `const` keyword added in Lua 5.4. In MTA Lua, the concept of a `const` is just that - a concept.*
 
-# Error handling
 
-TODO
+### Indentation and formatting
+
+Ensure your code editor (e.g. [Visual Studio Code](https://code.visualstudio.com/) 
+applies the rules established by the project's **.editorconfig** file.
 
 # Performance considerations
 
 -   Avoid unnecessary computations within loops.
 -   Cache results of expensive operations whenever possible.
 -   Use local variables to improve performance.
 
+### Don't use OOP functionality
+
+Enabling the [OOP](https://wiki.multitheftauto.com/wiki/OOP) functionality in any given resource will incur a performance hit. In general, pull requests to the official resources repository that enable it will generally fail code review.
+
+### Use range-based for loops rather than [`ipairs`](https://www.lua.org/manual/5.1/manual.html#pdf-ipairs)
+
+Range-based for loops are more performant than loops using `ipairs` and should be used whenever possible. For example:
+
+```lua
+-- Rather than this:
+for _, player in ipairs(getElementsByType("player")) do
+    iprint(player)
+end
+
+-- Do this:
+local players = getElementsByType("player")
+for i = 1, #players do
+    iprint(players[player])
+end
+```
```

Comment:
`ipairs` and `pairs` are much more readable than range-based loops. It doesnt make that much of a difference. Better to leave iterators as they are.

Reply:
It seems like most resources use ipairs/pairs, and as @TracerDS mentioned the difference in performance is so minuscule compared to a numeric for loop.

Instead, we should suggest using `ipairs` and `pairs` for readability reasons; unless it's a situation where it really matters for performance, which would be in rare cases.

Reply:
> `ipairs` and `pairs` are much more readable than range-based loops.

I do not see a readability difference between `ipairs` and range-based loops. They are one of the basic programming patterns that everyone learns.

> It seems like most resources use ipairs/pairs [...] the difference in performance is so minuscule compared to a numeric for loop.

It has been said that the difference [can be as high as 9%](https://github.com/multitheftauto/mtasa-resources/pull/363#discussion_r753804802). I think it's worth having in the performance considerations section, but I wouldn't deny a pull request for using `ipairs`. It's more of a guideline than a rule.

Reply:
On second review, in context of the performance considerations section, this is fine as a guideline.